worker/provisioner/host_preparer_test.go

```diff
@@ -0,0 +1,462 @@
+		// a) Forcibly restarting if Bridge() fails,
+		// b) Still observing and reporting our observation so that we at least
+		//    know what state we ended up in.
+		FuncName: "Release",
+	}})
+}
+
+func (s *hostPreparerSuite) TestPrepareHostObserveFailure(c *gc.C) {
+	s.Stub.SetErrors(
+		nil, // HostChangesForContainer
+		nil, // CreateBridger
+		nil, // AcquireLock
+		nil, // BridgeBridgeFailure
+		errors.New("cannot get network interfaces: enoent"), // GetObservedNetworkConfig
+	)
+	devices := []network.DeviceToBridge{{
+		DeviceName: "eth0",
+		BridgeName: "br-eth0",
+	}}
+	preparer := s.createPreparer(devices, nil)
+	containerTag := names.NewMachineTag("1/lxd/0")
+	logger := loggo.GetLogger("prepare-host.test")
+	err := preparer.Prepare(containerTag, logger)
+	c.Check(err, gc.ErrorMatches, `cannot discover observed network config: cannot get network interfaces: enoent`)
+	s.Stub.CheckCalls(c, []gitjujutesting.StubCall{{
+		FuncName: "HostChangesForContainer",
+		Args:     []interface{}{containerTag},
+	}, {
+		FuncName: "CreateBridger",
+	}, {
+		FuncName: "AcquireLock",
+	}, {
+		FuncName: "Bridge",
+		Args:     []interface{}{devices, 0},
+	}, {
+		FuncName: "ObserveNetwork",
+	}, {
+		// We don't SetHostMachineNetworkConfig, but we still release the lock
+		FuncName: "Release",
+	}})
+}
+
+func (s *hostPreparerSuite) TestPrepareHostObservedFailure(c *gc.C) {
+	s.Stub.SetErrors(
+		nil, // HostChangesForContainer
+		nil, // CreateBridger
+		nil, // AcquireLock
+		nil, // BridgeBridgeFailure
+		nil, // ObserveNetwork
+		errors.Unauthorizedf("failure"), // SetHostMachineNetworkConfig
+	)
+	devices := []network.DeviceToBridge{{
+		DeviceName: "eth0",
+		BridgeName: "br-eth0",
+	}}
+	preparer := s.createPreparer(devices, cannedObservedNetworkConfig)
+	containerTag := names.NewMachineTag("1/lxd/0")
+	logger := loggo.GetLogger("prepare-host.test")
+	err := preparer.Prepare(containerTag, logger)
+	c.Check(err, gc.ErrorMatches, `failure`)
+	s.Stub.CheckCalls(c, []gitjujutesting.StubCall{{
+		FuncName: "HostChangesForContainer",
+		Args:     []interface{}{containerTag},
+	}, {
+		FuncName: "CreateBridger",
+	}, {
+		FuncName: "AcquireLock",
+	}, {
+		FuncName: "Bridge",
+		Args:     []interface{}{devices, 0},
+	}, {
+		FuncName: "ObserveNetwork",
+	}, {
+		FuncName: "SetHostMachineNetworkConfig",
+		Args:     []interface{}{names.NewMachineTag("1"), cannedObservedNetworkConfig},
+	}, {
+		FuncName: "Release",
+	}})
+}
+
+func (s *hostPreparerSuite) TestPrepareHostCancel(c *gc.C) {
+	devices := []network.DeviceToBridge{{
+		DeviceName: "eth0",
+		BridgeName: "br-eth0",
+	}}
+	params := s.createPreparerParams(devices, nil)
+	lockName := params.LockName
+	ch := make(chan struct{})
+	close(ch)
+	params.AbortChan = ch
+	// This is what the AcquireLock should look like
+	params.AcquireLockFunc = func(abort <-chan struct{}) (mutex.Releaser, error) {
+		s.Stub.AddCall("AcquireLockFunc")
+		spec := mutex.Spec{
```

Comment:
:+1: I also added a test that the abort channel that we set up is being properly passed into the AcquireLock.